http.go

```diff
@@ -86,6 +109,86 @@ func (s *LumbermillServer) serveHealth(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusOK)
 }
 
+func (s *LumbermillServer) getHealthCheckClient(host string) (*influx.Client, error) {
+	s.healthCheckClientsLock.Lock()
+	defer s.healthCheckClientsLock.Unlock()
+
+	if client, exists := s.healthCheckClients[host]; !exists {
+		clientConfig := createInfluxDBClient(host, os.Getenv("INFLUXDB_SKIP_VERIFY") == "true")
+		client, err := influx.NewClient(&clientConfig)
+		if err != nil {
+			log.Printf("err=%q at=getHealthCheckClient host=%q", err, host)
+			return nil, err
+		} else {
+			s.healthCheckClients[host] = client
+			return client, nil
+		}
+	} else {
+		return client, nil
+	}
+}
+
+func (s *LumbermillServer) checkRecentTokens() map[string]int {
+	errantHosts := make(map[string]int)
+
+	s.recentTokensLock.RLock()
+	tokenMap := make(map[string]string)
+	for name, token := range s.recentTokens {
+		tokenMap[name] = token
+	}
+	s.recentTokensLock.RUnlock()
+
+	for name, token := range tokenMap {
+		client, err := s.getHealthCheckClient(name)
+		if err != nil {
+			errantHosts[name]++
+			continue
+		}
+
+		// Query the last point for the token across all queries and ensure it's been published to in the last XXX minutes.
+		for _, qfmt := range influxDbSeriesCheckQueries {
+			query := fmt.Sprintf(qfmt, token)
+
+			results, err := client.Query(query, influx.Second)
```

Comment:
Good point on the parallel queries. I'll make the update for that.

As for the timeouts, this will be hit by watchtower, not the ELB.
Presumably watchtower doesn't have a request timeout, but I'll have to
look..

On Mon, Apr 6, 2015 at 8:24 PM, Edward Muller <notifications@github.com>
wrote:

> In http.go
> <https://github.com/heroku/lumbermill/pull/17#discussion_r27843248>:
>
> > +		tokenMap[name] = token
> > +	}
> > +	s.recentTokensLock.RUnlock()
> > +
> > +	for name, token := range tokenMap {
> > +		client, err := s.getHealthCheckClient(name)
> > +		if err != nil {
> > +			errantHosts[name]++
> > +			continue
> > +		}
> > +
> > +		// Query the last point for the token across all queries and ensure it's been published to in the last XXX minutes.
> > +		for _, qfmt := range influxDbSeriesCheckQueries {
> > +			query := fmt.Sprintf(qfmt, token)
> > +
> > +			results, err := client.Query(query, influx.Second)
>
> Do you know what the timeout are on the client? They may exceed the
> timeouts for the health endpoint (behind and ELB).
>
> We also probably want to query all of these at in parrallel and gather the
> responses
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/heroku/lumbermill/pull/17/files#r27843248>.
>